actionpack/lib/action_controller/metal/redirecting.rb

```diff
@@ -92,10 +98,28 @@ def redirect_back(fallback_location:, allow_other_host: true, **args)
     #
     # All other options that can be passed to #redirect_to are accepted as
     # options and the behavior is identical.
-    def redirect_back_or_to(fallback_location, allow_other_host: true, **args)
-      referer = request.headers["Referer"]
-      redirect_to_referer = referer && (allow_other_host || _url_host_allowed?(referer))
-      redirect_to redirect_to_referer ? referer : fallback_location, **args
+    def redirect_back_or_to(fallback_location, allow_other_host: allow_open_redirects, **args)
+      referer = request.headers.fetch("Referer", fallback_location)
+      response_options = {
+        fallback_location: allow_open_redirects ? fallback_location : nil,
+        allow_other_host: allow_other_host,
+        **args,
+      }
+      redirect_to referer, response_options
+    end
+
+    def _compute_safe_redirect_to_location(request, options, response_options)
+      location = _compute_redirect_to_location(request, options)
+      if response_options[:allow_other_host] || _url_host_allowed?(location)
+        location
+      else
+        fallback_location = response_options[:fallback_location] || raise(ArgumentError, <<~MSG.squish)
+          Unsafe redirect #{location.inspect},
```

Comment:
This PR feels like it's in part to allow redirects via params, but if we're injecting that here could an attacker fill a buffer and take down the process? Perhaps NGINX would block this via some config, but some apps just run Puma directly without it.

Reply:
Hm, I hadn't considered that. If I understand correctly, truncating the location might help since the location at this should always be a string.

Reply:
Could try using `truncate`, yeah.